src/storage/mvcc/reader.rs

```diff
@@ -417,4 +418,236 @@ impl<'a> MvccReader<'a> {
+        assert_eq!(reader.need_gc(safe_point), need_gc);
+    }
+
+    #[test]
+    fn test_need_gc() {
+        let path = TempDir::new("_test_storage_mvcc_reader").expect("");
+        let path = path.path().to_str().unwrap();
+
+        let region = make_region(1, vec![0], vec![10]);
+
+        // Open without properties.
+        {
+            let db = open_db(path, false);
+            let mut engine = RegionEngine::new(db.clone(), region.clone());
+
+            // Put 2 keys.
+            engine.put(&[1], 1, 1);
+            engine.put(&[4], 2, 2);
+            check_need_gc(db.clone(), region.clone(), 10, true);
+            engine.flush();
+            // After this flush, we have a SST file without properties.
+            // Without properties, we always need GC.
+            check_need_gc(db.clone(), region.clone(), 10, true);
+        }
+
+        // Open with properties.
+        {
+            let db = open_db(path, true);
+            let mut engine = RegionEngine::new(db.clone(), region.clone());
+
+            // Put 2 keys.
+            engine.put(&[2], 3, 3);
+            engine.put(&[3], 4, 4);
+            engine.flush();
+            // After this flush, we have a SST file w/ properties, plus the SST
+            // file w/o properties from previous flush. We always need GC as
+            // long as we can't get properties from any SST files.
+            check_need_gc(db.clone(), region.clone(), 10, true);
+            engine.compact();
+            // After this compact, the two SST files are compacted into a new
+            // SST file with properties. Now all SST files have properties and
+            // all keys have only one version, so we don't need gc.
+            check_need_gc(db.clone(), region.clone(), 10, false);
+
+            // Put 2 more keys and delete them.
+            engine.put(&[5], 5, 5);
+            engine.put(&[6], 6, 6);
+            engine.delete(&[5], 7, 7);
+            engine.delete(&[6], 8, 8);
+            engine.flush();
+            // After this flush, keys 5,6 in the new SST file have more than one
+            // versions, so we need gc.
+            check_need_gc(db.clone(), region.clone(), 10, true);
+            // But if the `safe_point` is older than all versions, we don't need gc too.
+            check_need_gc(db.clone(), region.clone(), 0, false);
+
+            // We gc the two deleted keys manually.
+            engine.gc(&[5], 10);
+            engine.gc(&[6], 10);
+            engine.compact();
+            // After this compact, all versions of keys 5,6 are deleted,
+            // no keys have more than one versions, so we don't need gc.
+            check_need_gc(db.clone(), region.clone(), 10, false);
+
+            // A single lock version need gc.
+            engine.lock(&[7], 9, 9);
+            engine.flush();
+            check_need_gc(db.clone(), region.clone(), 10, true);
+        }
+    }
```

Comment:
Could you split this test to `test_gc_without_db_properties ` and `test_gc_with_db_properties`?